cleaner/cleaner_main.py

```diff
@@ -0,0 +1,65 @@
+# Copyright (c) Microsoft Corporation
+# All rights reserved.
+#
+# MIT License
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+# to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+# BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import time
+from datetime import timedelta
+from cleaner.scripts import clean_docker_cache
+from cleaner.worker import Worker
+from cleaner.utils.logger import LoggerMixin
+from cleaner.utils import common
+
+
+class Cleaner(LoggerMixin):
+
+    def __init__(self):
+        self.workers = {}
+
+    def add_worker(self, key, worker):
+        if key not in self.workers:
+            self.workers[key] = worker
+        else:
+            self.logger.warn("worker with key %s already exists.", key)
+
+    def start(self):
+        for k, w in self.workers.items():
+            w.start()
+            self.logger.info("worker %s started.", k)
+
+    def sync(self):
+        try:
+            for w in self.workers.values():
+                w.join()
+        except:
+            self.logger.error("cleaner interrupted and will exit.")
+            for w in self.workers.values():
+                w.terminate()
```

Comment:
>w.terminate() [](start = 16, length = 13)

Suggest to add try/except here also so that if some worker's terminate() failed, we can continue on to the next.

Reply:
add a new method cleaner.terminate() to kill a worker and all its subprocesses.
This method will catch all exceptions when killing each worker.